### PR TITLE
[SPARK-11322][PySpark] Keep full stack trace in captured exception

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-11322

As reported by @JoshRosen in [databricks/spark-redshift/issues/89](https://github.com/databricks/spark-redshift/issues/89#issuecomment-149828308), the exception-masking behavior sometimes makes debugging harder. To deal with this issue, we should keep full stack trace in the captured exception.

### CODE REVIEW

**Comment by viirya** [comment]
ping @davies any comments about this? Is it in right direction?

**Comment by davies** [comment]
@viirya Not sure, could you post an example how the full stack trace looks like?

**Comment by viirya** [comment]
When the exception is thrown from JVM, we see a simplified description:

    >>> df.select(sha2(df.a, 1024)).collect()
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/home/hduser/repos/spark/python/pyspark/sql/functions.py", line 943, in sha2
        jc = sc._jvm.functions.sha2(_to_java_column(col), numBits)
      File "/home/hduser/repos/spark/python/lib/py4j-0.9-src.zip/py4j/java_gateway.py", line 813, in __call__
      File "/home/hduser/repos/spark/python/pyspark/sql/utils.py", line 52, in deco
        raise IllegalArgumentException(s.split(': ', 1)[1], stackTrace)
    pyspark.sql.utils.IllegalArgumentException: u'requirement failed: numBits 1024 is not in the permitted     values (0, 224, 256, 384, 512)'

If we catch the exception, we can print its full stack trace:

    >>> from pyspark.sql.utils import AnalysisException, IllegalArgumentException
    >>> try:
    ...     df.select(sha2(df.a, 1024)).collect()
    ... except IllegalArgumentException as e:
    ...     print e.stackTrace
    ... 
    scala.Predef$.require(Predef.scala:233)
    org.apache.spark.sql.functions$.sha2(functions.scala:1596)
    org.apache.spark.sql.functions.sha2(functions.scala)
    sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    java.lang.reflect.Method.invoke(Method.java:483)
    py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:231)
    py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:381)
    py4j.Gateway.invoke(Gateway.java:259)
    py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)
    py4j.commands.CallCommand.execute(CallCommand.java:79)
    py4j.GatewayConnection.run(GatewayConnection.java:209)
    java.lang.Thread.run(Thread.java:745)

File: python/pyspark/sql/utils.py [hunk 1]
```diff
@@ -36,10 +45,11 @@ def deco(*a, **kw):
             return f(*a, **kw)
         except py4j.protocol.Py4JJavaError as e:
             s = e.java_exception.toString()
+            stackTrace = '\n'.join(map(lambda x: x.toString(), e.java_exception.getStackTrace()))
```
Comment (davies) [inline python/pyspark/sql/utils.py]: '\n\t at ', make it looks like a Java stacktrace

File: python/pyspark/sql/utils.py [hunk 2]
```diff
@@ -36,10 +45,12 @@ def deco(*a, **kw):
             return f(*a, **kw)
         except py4j.protocol.Py4JJavaError as e:
             s = e.java_exception.toString()
+            stackTrace = '\n\t at'.join(map(lambda x: x.toString(),
```
Comment (davies) [inline python/pyspark/sql/utils.py]: another space after `at`?

Comment (viirya) [inline python/pyspark/sql/utils.py]: yap.

**Comment by viirya** [comment]
@davies Thanks. I think this is ready now.

**Pull request closed without merging**